koules.sdl.6
.TH KOULES 6 "9 nov 2015" "Linux" "games"

.SH NAME
koules.sdl \- Action game for linux

.SH SYNOPSIS
koules.sdl [-slMmdh]

.SH KOULES FROM THE TOP:
.br

.br
                             .
.br
                 ---       .  .                    O
.br
                /   \\        . .
.br
               |     |    . . .. ---
.br
                \\   /           /   \\        X           O
.br
                 ---           |    O|
.br
                                \\  O/
.br
                                 ---       ---
.br
                .       O                 /O O\\
.br
             . . . .                     |     |
.br
              . . .                       \\   /
.br
            .  .X.  .          O           ---
.br
              . . .                         .
.br
             . . . .                       ...
.br
                .                         . . .
.br
                                         . . . .
.br
                                        .   .   .
.br
                        H A V E
.br
                          FUN
.br
                           !


.SH DESCRIPTION
.B Koules
is an action game. The mission is to bounce all koules fighters from your 
sector

.SH "OPTIONS"
.TP
.B \-h 
for help
.TP
.B \-s 
for small display(320x200)
for worse machines
.TP
.B \-l 
for large display(640x480)
default size
.TP
.B \-M 
disable mouse support
.TP
.B \-d 
Disable sound support
Game will run faster.
.SH "Network options"
.TP
.B \-C<host>
run koules as network client
.TP
.B \-S
run koules as network server
.TP
.B \-P<port>
select port. Default is:12345
    Use this in case that some other program is already using this port.

.SH "Server options"
.TP
.B \-E
run server in extra large-support for 800x620 X clients.
.TP
.B \-W
run server in wide mode-support for 320x200 svgalib and OS/2 clients.
    Use this in case that server is refusing your clients.
.TP
.B \-L<level>
select level for server
.TP
.B \-D<number>
select dificulty for server:
     0: nightmare
     1: hard
     2: medium(default and recomended)
     3: easy
     4: very easy
.TP
.B \-K
run server in deathmatch mode
.SH "Points:"
.br
  Bounce something         :  +1
.br
  Eat extension            :  +10
.br
  Eat thief		   :  -30
.br
  Eat friend		   :  +30
.br
.B  At end of level
.br
  every live               :  +20
.br
  every lost level         :  -100
.br
.B  In deathmatch mode
.br
  every destroyed concurrent:  +100


.SH PORTABILITY

The program is written in ANSI C and uses the SDL graphics library.

.SH BUGS

Probably not as fast as it could be.

Doesn't use screen area wisely.

.SH SEE ALSO
http://www.paru.cas.cz/~hubicka/koules/English/koules.html
.SH AUTHOR
Jan Hubicka

.B email:
hubicka@limax.paru.cas.cz
.br

.B smail:
.br
       Jan Hubicka
.br
       Dukelskych bojovniku 1944
.br
       Tabor 39001
.br
       Czech Republic



ChangeLog
Version 1.4
	o	Better joystick support by Ludvik Tesar
        o	'h' for help
	o	Fixes problems with non USS-lite(voxware) sound drivers
		known to work well with ultrasoud project
	o	Fixes for svgalib-1.2.11
	o	Fixed some other small problems
	o	Fixed bug in X11 and intro texts
	o	Be more robust for old and/or incorrect rc files
	o	Fixed security bug in SVGAlib version
	o	Socklib complettely rewrote to sock.c to avoid copyright
		problems with commercial use.
		Sock solves some old problems with network code, but may
		introduce new ones. Please let me know
---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------
Version 1.3	(Bugfix..)
	o	Fixed crash in 51. level

---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------
Version 1.2	(SGI sound server)
Major changes
	o	Sound server for SGI by Rick Sayre
	o	Sound server moved from koules directory to sound directory
Minor changes
	(patches by Rick Sayre)
	o	new creator effect
	o	SGI sound server
	o	Fixed socklib for compatibility with SGI
	o	spelling check
	(patches by Jan Hubicka)
	o	bit improved Rick's creator effect
	o	new socklib version
	o	probably fixed random crash bug that happended on SGI
	o	Nasty network code bug found and fixed...no more
		connection refused and similiar error messages
		(I hope :)
	o	Fixed bug with byte-ordering in soularis sound server
	o	replaced mkdir by mkdirhier in Imakefile
	o	Added missing errno.h into server.c and client.c
	o	Fixed possible owerflowing in rotation of ship at
		network play
	o	Fixed bug in outro with more players
	o	Added historic directory
	o	removed calculatin of bitmap for 6th badie when
		actually there is only 5 of them(and of course all
		arrays are alocated just for five :)
	o	Added experimental page flipping mode
	o	Iproved text.h
---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------
Version 1.2beta1	(Networked version,OS/2 port)
Major Changes
	o	Four new creatures by Kamil Toman, Tomas Marsh and me!
	o	Deathmatch for one player! (Is there some another game
		with this feature)
	o	OS/2 Warp support by Thomas A. K. Kjaer!
	o	Network support! (tested on ethernet and plip line)
	o	Official koules WWW homepagepage 
                with documentation, screenshots, tips and tricks etc..

     >>>URL=http://www.paru.cas.cz/~hubicka/koules/English<<<

	o	Added new extra large screen resolution..(experimental)
		by Bertrand Leroy and rewrote by me..
	o	Changed some command line options..see manual pages
	o	Improved deathmatch by Bertrand Leroy and rewrote by me
	o	Added new user friendly starting dialog for RedHat!
	o	Use shared images - no shared pixmaps! great for
		older and non Xfree servers! Now MITSHM mode is possible
		on most X's
	o	Mitshm support now works on 16,24 and 32 bpp too.
Minor Changes
	o	Better private colormap support-keep first 20 colors...
		so no more blinking of windows...
	o	Renamed usleep emulation function to myusleep so
		it does not collide with normal usleep so usleep
		emulation should be enabled on platforms with usleep
		too
	o	Added joystick.h into distribution to allow compile
		koules joystick support w/o joystick toolkid installed.
		good for building rpms etc...
	o	Improved mitshm detection-game will not crash when 
		mitshm is not possible
	o	drawing of letters in mitshm mode too.
	o	Fixed compatibility for DEC
	o	Added gameplan into rc files
	o	Removed requirements part from manual pages
	o	Added mygetopt for better comtability
	o	Added level display at bottom of screen
	o	Added switching of console when koules is started from X
	o	ESC in submenus won't exit game anymore...
	o	Deathmatch changes:
		o	New scoring..
		o	Allow spitting
		o	Holes and inspector at the beggining
		o	Allow new holes and bbals at end
		o	Spitting between ships
		o	Added GET READY message
	o	Removed useless wasmouse variable
	o	Added draw_reversed ball bitmap for drawing thiefs and goodies 
	o	Added THIEFS that eats all extensions (Kamil Toman and me)
	o	Added LUNATICS that do something very strange..(Tomas Marsh and me)
	o	Added FINDERS that brings lots of extensins (Kamil Toman and me)
	o	Added TTOOL (Kamil Toman and me)
	o	Added some of Bertrand Leroy's patches
	o	Added Thomas March's speling csechks
	o	Added Thomas T. K. Kjaer's patches into official linie
	o	Improved sounds in gameplan code
	o	Updated manual pages and readme
	o	Removed question about NETWORK support in FAQ
	o	Added network oriented switches: SCPWDKL
	o	Updated all makefiles
	o	Better error signal handling in svgalib version
	o	Fixed some small and innocent bugs...
	o	Added support for 640x380 mode (wide mode)
		to make possible combine svga 320x200 clients with
		svgalib 640x480 and X's ones.
	o	-S (Sync) renamed to -y because of -S (server) conflict
	o	fixed unused variable c warning in framebuffer.c
	o	added compile time flag NETSUPPORT
	o	created client.c
	o	created server.c
	o	created socklib.c
	o	created objectio.c
	o	network releated changes: 
		o	to menu.c added client's menus
		o	to koules.c added game parts of client and server loop,
			handling of keys devided into two parts-server and 
			client, added sending of events to pyrotechnic handling
			functions
		o	gameplan.c
			process effects at clients and possitions at server :)
		o	xlib/init.c and svgalib/init.c added -C and -S options
			for server and client
---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------- 8< ---------
Version 1.1c 10/1 95	(Saving of rc files, improved portability)
Major changes
	o	removed gl_allocatecontext for svgalib1.2.0 compatibility
	o	levels selector
	o	added rc files
	o	fast unrolled assembler memcpy for i386/mitshm mode
	o	fixed for DEC
	o	fixed for AIX
	o	fixed for solaris w/o gcc
	o	fixed for IRIX
	o	experimental rplay support
	o	monochrome and black and white displays support
	o	dithering
	o	HP-UX sound server by Lutz Vieweg <lkv@mania.robin.de>
Minor Changes
	o	fixed for svgalib-1.2.8

koules.tcl
#!/usr/bin/wish
wm title . "Koules 1.4"
# This is simple user (anti)friendly dialog for starting koules
# Only reason why I did it is that I wanted to try tcl/tk
#files - configure here if paths are differ
set xkoules "/usr/bin/X11/xkoules"
set koulessvga "/usr/local/bin/koules.svga"
set koulessound "/usr/local/lib/koules/creator1.raw"
set config "~/.xkoules.opt"
set revision 1.1
# Set default
set GameMode Single
set Shm 1
set private 0
set Fullscreen 0
set ScreenSize "-l"
set Sound 1
set Monochrome 0
set blackwhite 0
set Cursor 0
set Level 1
set Host localhost
set Deatmatch 0
set SPort 12345
set CPort 12345
set Difficulty 2
set SScreenSize -l
if {[file exists $config] == 1} {
  set FileId [open $config "r"]
  if {[gets $FileId] == $revision } {
  set GameMode [gets $FileId]
  set ScreenSize [gets $FileId ]
  set private [gets $FileId ]
  set blackwhite [gets $FileId ]
  set Cursor [gets $FileId ]
  set Shm [gets $FileId ]
  set Monochrome [gets $FileId ]
  set Sound [gets $FileId ]
  set Fullscreen [gets $FileId ]
  set CPort [gets $FileId ]
  set Host [gets $FileId ]
  set SPort [gets $FileId ]
  set Level [gets $FileId ]
  set Difficulty [gets $FileId ]
  set Deathmatch [gets $FileId ]
  set SScreenSize [gets $FileId ]
  } else {puts "Ignoring config file...another version"}
  close $FileId
}

frame .gamemode -relief raised -borderwidth 3
radiobutton .gamemode.single \
   -value "Single"\
   -variable GameMode\
   -text "Single"\
   -command "SetMode"
radiobutton .gamemode.client \
   -value "Client"\
   -variable GameMode\
   -command "SetMode" \
   -text "Client"
radiobutton .gamemode.server \
   -value "Server"\
   -variable GameMode\
   -command "SetMode"\
   -text "Server"

pack .gamemode.single\
     .gamemode.client\
     .gamemode.server\
   -side left -expand yes
frame .buttons -relief raised
frame .single -relief raised
frame .single.size -relief raised
frame .single.gameopt -relief raised
frame .server -relief raised
frame .server.left -relief raised
frame .server.left.size -relief raised
frame .server.left.data -relief raised
frame .server.text -relief raised
frame .server.difficulty -relief raised

radiobutton .server.difficulty.0 \
   -value "0"\
   -variable Difficulty\
   -text "Nightmare"
radiobutton .server.difficulty.1 \
   -value "1"\
   -variable Difficulty\
   -text "Hard"
radiobutton .server.difficulty.2 \
   -value "2"\
   -variable Difficulty\
   -text "Medium"
radiobutton .server.difficulty.3 \
   -value "3"\
   -variable Difficulty\
   -text "Easy"
radiobutton .server.difficulty.4 \
   -value "4"\
   -variable Difficulty\
   -text "Very easy"
radiobutton .server.left.size.s \
   -value "-l"\
   -variable SScreenSize\
   -text "350x200 and 640x480"
radiobutton .server.left.size.m \
   -value "-W"\
   -variable SScreenSize\
   -text "320x200 and 640x380"
radiobutton .server.left.size.l \
   -value "-E"\
   -variable SScreenSize\
   -text "900x600" 
checkbutton .server.left.size.deathmatch \
   -variable Deathmatch\
   -text "Deathmatch"\
   -command "SetDeath"
frame .server.left.data.level
frame .server.left.data.port
label .server.left.data.level.label -text Level
label .server.left.data.port.label -text Port
entry .server.left.data.level.entry -relief sunken -width 3 -textvariable Level
#.server.left.data.level.entry insert 0 $Level
entry .server.left.data.port.entry -relief sunken -width 6 -textvariable SPort
#.server.left.data.port.entry insert 0 $SPort 
frame .client
label .client.hostlabel -text Host
label .client.portlabel -text Port
entry .client.host -relief sunken -textvariable Host
entry .client.port -relief sunken -width 6 -textvariable CPort
#.client.host insert 0 $Host
#.client.port insert 0 $CPort
pack .client.hostlabel\
     .client.host\
     .client.portlabel\
     .client.port -side left -fill x
#server.gameopt.level insert $Level
pack .server.left.data.level.label\
     .server.left.data.level.entry -side left -fill x
pack .server.left.data.port.label\
     .server.left.data.port.entry -side left -fill x
pack .server.left.data.port\
     .server.left.data.level -side top -fill x
pack .server.difficulty.0\
     .server.difficulty.1\
     .server.difficulty.2\
     .server.difficulty.3\
     .server.difficulty.4\
     -padx 10 -expand 1 -anchor w
pack .server.left.size.s\
     .server.left.size.m\
     .server.left.size.l \
     .server.left.size.deathmatch -padx 10 -expand 1 -anchor w
pack .server.left.data\
     .server.left.size
#normal game options
radiobutton .single.size.s \
   -value "-s"\
   -variable ScreenSize\
   -text "320x250"
radiobutton .single.size.m \
   -value "-l"\
   -variable ScreenSize\
   -text "640x480"
radiobutton .single.size.l \
   -value "-E"\
   -variable ScreenSize\
   -text "900x600" 
checkbutton .single.size.mode \
   -variable Fullscreen\
   -text "Fullscreen"\
   -command "SetExtra"
checkbutton .single.gameopt.sound\
   -variable Sound\
   -text "sound"
checkbutton .single.gameopt.shm\
   -variable Shm\
   -text "Use Shared memory"
checkbutton .single.gameopt.monochrome\
   -variable Monochrome\
   -text "Monochrome"
checkbutton .single.gameopt.blackwhite\
   -variable blackwhite\
   -text "Black and white" 
checkbutton .single.gameopt.private\
   -variable private\
   -text "Private colormap"
checkbutton .single.gameopt.cursor\
   -variable Cursor\
   -text "Game own mouse cursor"
button .buttons.start -text Start -command "Start"
button .buttons.quit -text Quit -command exit
pack .buttons.start\
     .buttons.quit\
     -expand 1 -fill x -side left
pack .single.size.s\
     .single.size.m\
     .single.size.l\
     .single.size.mode -padx 10 -fill x -expand y -anchor w
pack .single.gameopt.sound \
     .single.gameopt.shm \
     .single.gameopt.private \
     .single.gameopt.monochrome \
     .single.gameopt.blackwhite \
     .single.gameopt.cursor  -side top  -anchor w
pack .single.size\
     .single.gameopt -side left
pack .server.left .server.difficulty -side left -fill x
pack .gamemode -side top -expand y -fill x -pady 2m
pack .single  -expand y -fill x -pady 2m
pack .buttons -side bottom -expand y -fill x -pady 2m
if {[file exists $xkoules] == 1} {
  set Koulesexe $xkoules
} else {
   if {[file exists $koulessvga] == 1} {
     set Fullscreen 1
     .single.size.mode configure -state disabled
   } else {
     puts "Koules executable not found"
     exit
   }
}
if {[file exists $koulessound] == 0} {
  set Sound 0
  .single.gameopt.sound configure -state disabled
}
if {[file exists $koulessvga] == 1} {
  set Koulesexe $koulessvga
} else {
  set Fullscreen 0
  .single.size.mode configure -state disabled
}
proc SetDeath {} {
  global Deathmatch
  if {$Deathmatch == "1" } {
    .server.left.data.level.entry configure -state disabled
  } else {
    .server.left.data.level.entry configure -state normal
  }
}
SetDeath
proc SetExtra {} {
  global Fullscreen
  global ScreenSize
  if {$Fullscreen == "1" } {
     .single.size.l configure -state disabled
     .single.size.s configure -text 320x200
     .single.gameopt.shm configure -state disabled
     .single.gameopt.private configure -state disabled
     .single.gameopt.monochrome configure -state disabled
     .single.gameopt.blackwhite configure -state disabled
     .single.gameopt.cursor configure -state disabled
     if {$ScreenSize == "-E"} {
       set ScreenSize "-l"
       #.single.size.m configure -state active
     }
  } else {
     .single.size.l configure -state normal
     .single.size.s configure -text 320x250
     .single.gameopt.shm configure -state normal
     .single.gameopt.private configure -state normal
     .single.gameopt.monochrome configure -state normal
     .single.gameopt.blackwhite configure -state normal
     .single.gameopt.cursor configure -state normal
  }
}
SetExtra
proc Start {} {
  global GameMode
  global ScreenSize
  global private
  global blackwhite
  global Cursor
  global Shm
  global Monochrome
  global Sound
  global Fullscreen
  global config
  global revision
  #client options
  global CPort
  global Host
  #server options
  global SPort
  global Level
  global Difficulty
  global Deathmatch
  global SScreenSize
  global xkoules
  global koulessvga
  global Koulesexe
  set FileId [open $config "w"]
  puts $FileId $revision
  puts $FileId $GameMode
  puts $FileId $ScreenSize
  puts $FileId $private
  puts $FileId $blackwhite
  puts $FileId $Cursor
  puts $FileId $Shm
  puts $FileId $Monochrome
  puts $FileId $Sound
  puts $FileId $Fullscreen
  puts $FileId $CPort
  puts $FileId $Host
  puts $FileId $SPort
  puts $FileId $Level
  puts $FileId $Difficulty
  puts $FileId $Deathmatch
  puts $FileId $SScreenSize
  close $FileId
  if {$GameMode != "Server" } {
    set opt $ScreenSize
    if {$Sound == "0"} {
      set opt "$opt -d"
    }
    if {$Fullscreen == "0"} {
    set exec $xkoules
    if {$private == "1"} {
      set opt "$opt -p"
    }
    if {$Monochrome == "1"} {
      set opt "$opt -m"
    }
    if {$blackwhite == "1"} {
      set opt "$opt -b"
    }
    if {$Cursor == "1"} {
      set opt "$opt -x"
    }
    if {$Shm == "0"} {
      set opt "$opt -M"
    }
    } else {
      set exec $koulessvga
    }
    if {$GameMode=="Client"} {
      set opt "$opt -C $Host -P $CPort"
    }
    #puts "$exec $opt"
    destroy .
    exec /bin/sh "-c" "$exec $opt"
  } else {
    set exec  $Koulesexe
    set opt "-S -P $SPort $SScreenSize -D $Difficulty"
    if {$Deathmatch == "1"} {
      set opt "$opt -K"
    }
    #puts "$exec $opt"
    destroy .
    exec xterm "-iconic" "-title" "Koules server" "-e" "/bin/sh" "-c" "$exec $opt"
  }
}
pack .gamemode\
     .buttons -expand 1 -side top -pady 2 -anchor w
proc SetMode {} {
  global GameMode
  if {$GameMode == "Single" } {
    pack forget .server
    pack forget .client
    pack .single\
         -before .buttons -expand 1 -side top -pady 2 -anchor w
  }
  if {$GameMode == "Server"} {
    pack forget .single
    pack forget .client
    pack .server -before .buttons -expand 1 -side top -pady 2 -anchor w
  }
  if {$GameMode == "Client"} {
    pack forget .server
    pack .single\
         -before .buttons -expand 1 -side top -pady 2 -anchor w
    pack .client\
         -before .buttons -expand 1 -side top -pady 2 -anchor w
  }
}
puts $GameMode
bind . <Return> Start
bind . <Escape> exit
bind . <Control-c> exit
SetMode
